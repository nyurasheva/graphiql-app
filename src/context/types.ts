export type Translation = {
  [key: string]: string;
};

export type Translations = {
  [key: string]: Translation;
};

export type LocalizationContextType = {
  language: string;
  translations: Translation;
  changeLanguage: (newLanguage: string) => void;
};
