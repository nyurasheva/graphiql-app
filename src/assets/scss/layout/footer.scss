/* footer */

.footer {
  flex-shrink: 0;
  @include font($family: $font-family-title);

  &__content {
    @include flex($justify: space-between, $wrap: wrap);
    padding: 1em 0;
    gap: 10px 40px;
    color: $background;

    @include media('s') {
      gap: 10px;
    }

    & > * {
      flex: 1;
    }
  }

  &__svg {
    height: 3rem;
    fill: $background;
    transition: 0.25s;
    filter: invert(100%);

    &:hover {
      scale: 1.1;
    }
  }

  &__rss-link {
    text-align: left;

    svg {
      background: url('../svg/rs.svg') center / contain no-repeat;
      width: 8rem;

      @include media('s') {
        width: 5rem;
      }
    }
  }

  &__gh {
    @include flex();
    gap: 1rem;

    &-link svg {
      background: url('../svg/github.svg') center / contain no-repeat;
      width: 3rem;

      @include media('s') {
        width: 2rem;
      }
    }
  }

  &__year-link {
    text-align: right;

    @include media('s') {
      font-size: 0.8rem;
    }
  }
}
