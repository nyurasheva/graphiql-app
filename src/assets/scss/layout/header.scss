/* header */

.header {
  position: sticky;
  @include flex();
  width: 100%;
  top: 0;
  left: 0;
  right: 0;
  z-index: 10;
  height: 8rem;
  background: $background-dark;
  transition: $transition;

  &.sticky {
    background: #1e1e1e;
  }

  &__content {
    @include flex($justify: space-between);
    gap: 20px;
    padding: 0 20px;
  }

  @include media('s') {
    height: 5rem;
    flex-wrap: wrap;
    justify-content: end;
    padding: 0 10px;
    gap: 10px;
  }

  @include media('xxs') {
    gap: 0;
  }

  & > * {
    flex: 1;
  }

  &__logo {
    text-align: left;

    .logo {
      padding: 15px 0;
      min-width: 150px;
      max-width: 280px;
      position: relative;
      margin-top: 2px;

      @include media('xs') {
        min-width: auto;
      }

      a {
        white-space: nowrap;
      }

      &__image {
        display: inline-block;
        max-height: 50px;
        max-width: 100%;
        width: auto;
        text-decoration: none;
        margin-top: -22px;

        @include media('xs') {
          max-height: 40px;
        }

        @include media('xxs') {
          max-height: 30px;
        }
      }

      &__span {
        @include font(
          $family: $font-family-title,
          $size: 1.6rem,
          $weight: bold,
          $color: $links
        );
        margin-left: 10px;

        @include media('xs') {
          display: none;
        }
      }
    }
  }

  &__links {
    @include flex($justify: center);
    gap: 30px;
  }

  &__buttons {
    @include flex($justify: flex-end);
    gap: 10px;

    @include media('s') {
      flex: 0;
    }

    @include media('xxs') {
      flex: 1;
    }

    &-icon {
      @include media('s') {
        @include flex($justify: flex-start, $align: baseline);
        position: relative;
        left: 0;
        top: 0;
        right: 0;
        background: url(/src/assets/svg/profile.svg) no-repeat center center /
          cover;
        cursor: pointer;
        height: 25px;
        width: 25px;
      }
    }

    &-list {
      @include flex($justify: flex-end);
      gap: 10px;

      @include media('s') {
        display: none;
      }

      &.active {
        @include media('s') {
          position: absolute;
          @include flexColumn($align: flex-start);
          gap: 10px;
          margin-top: 133px;
          margin-right: 42px;
          background: #fad7ee;
          padding: 7px 14px;
          border-radius: 12px;
          box-shadow:
            0 0 0 1px #ffffff0a,
            0 5px 7px 0 #ffffff38;
          z-index: 1;
        }

        &::before,
        &::after {
          content: '';
          position: absolute;
          top: -12px;
          right: 7%;
          border-color: transparent;
          border-style: solid;
          border-width: 6.5px;
          border-bottom-color: #fad7ee;
        }

        &::before {
          top: -13px;
          border-bottom-color: rgba(0, 0, 0, 0.1);
        }
      }
    }
  }
}
