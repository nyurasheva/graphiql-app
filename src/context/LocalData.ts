import { Translations } from './types';

// Simulated translations data
export const translationsData: Translations = {
  en: {
    en: 'En',
    ru: '',
    welcome: 'Welcome',
    main: 'Editor',
    registration: 'Registration',
    login: 'Sign in',
    logout: 'Sign out',
    registerTitle: 'Create an account',
    registerSubtitle: 'Registered user benefit:',
    registerText: 'Already have an account? Then just click ',
    loginTitle: 'Sign in',
    loginSubtitle: 'Not registered?',
    loginText: 'Creating an account will take no more than a minute ',
    notfoundText: 'Error',
    notfoundTitle: 'Sorry! We could not find what you were looking for. ',
    notfoundInfo: 'The page was not found.',
    notfoundLink: 'Go to home page',
    submit: 'Submit',
    api: 'API',
    headers: 'Headers',
    variables: 'Variables',
    request: 'Request',
    response: 'Response',
    graphqlTitle: ' — this is the ability to harness all the power',
    graphqlText: 'make a request into your own universe',
    graphqlDescr:
      'The task is to implement a playground/IDE for working with GraphQL queries. GraphQL is a query language used by client applications to handle data. GraphQL is associated with the concept of a schema that allows organizing the creation, reading, updating, and deleting of data in the application. GraphiQL is an open-source tool. In addition to the IDE itself for working with GraphQL queries, authorization capabilities are implemented to grant access to the tool only to authorized users. ',
    dogTitle: `Let's do something together!`,
    dogText1: 'Try to use our editor created thanks to the course ',
    dogText2:
      'By taking courses, you get: free training, technological materials, a mentor, new friends, a certificate and much more.',
    dogBtn: 'Go to the editor',
    aboutTitle: 'You get it!',
    plusTitle: 'Advantages',
    plusBtn: 'Go to the editor',
    mainTitle: 'GraphiQL Editor',
    attentionApi: 'Attention! Submitting API...',
    attentionHeaders: 'Attention! Submitting Headers...',
    attentionVariables: 'Attention! Submitting Variables...',
    docsLink: 'Docs',
    ourTeam: 'Our team',
  },
  ru: {
    en: '',
    ru: 'Рус',
    welcome: 'Приветствие',
    main: 'Редактор',
    registration: 'Регистрация',
    login: 'Вход',
    logout: 'Выход',
    registerTitle: 'Создать учетную запись',
    registerSubtitle: 'Преимущества зарегистрированного пользователя:',
    registerText: 'Уже есть учетная запись? Тогда просто жмите ',
    loginTitle: 'Вход',
    loginSubtitle: 'Не зарегистрированы?',
    loginText: 'Создание учетной записи займет не больше минуты. ',
    notfoundText: 'Ошибка',
    notfoundTitle: 'Извините! Мы не смогли найти то, что вы искали.',
    notfoundInfo: 'Запрашиваемая страница не найдена.',
    notfoundLink: 'Перейти на главную страницу',
    submit: 'Отправить',
    api: 'АПИ',
    headers: 'Заголовок запроса',
    variables: 'Переменные',
    request: 'Запрос',
    response: 'Ответ',
    graphqlTitle: ' — это возможность использовать всю мощь',
    graphqlText: 'сделай запрос в свою собственную вселенную',
    graphqlDescr:
      'Задача состоит в том, чтобы реализовать playground/IDE для работы с запросами GraphQL.GraphQL - это язык запросов, используемый клиентскими приложениями для обработки данных.GraphQL ассоциируется с таким понятием, как schema, которое позволяет организовать создание, чтение, обновление и удаление данных в приложении. GraphiQL - это инструмент с открытым исходным кодом.Помимо самого ide для работы с запросами GraphQL, реализованы возможности авторизации, чтобы предоставить доступ к инструменту только авторизованным пользователям. ',
    dogTitle: 'Давайте делать что-нибудь вместе!',
    dogText1:
      'Попробуйте в использовании наш редактор, созданный благодаря курсу от ',
    dogText2:
      'Обучаясь на курсе вы получаете: бесплатное обучение, учебные материалы, ментора, новых друзей, сертификат и многое другое.',
    dogBtn: 'Перейти в редактор',
    aboutTitle: 'Вы получаете!',
    plusTitle: 'Преимущества',
    plusBtn: 'Перейти в редактор',
    mainTitle: 'GraphiQL Редактор',
    attentionApi: 'Внимание! Отправка API...',
    attentionHeaders: 'Внимание! Отправка заголовков...',
    attentionVariables: 'Внимание! Отправка переменных...',
    docsLink: 'Схема',
    ourTeam: 'Наша команда',
  },
};
