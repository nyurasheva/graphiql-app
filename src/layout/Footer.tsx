import { developersEn } from '../constants/developers';
import { WELCOME_ROUTE } from '../constants/route';

const Footer = () => {
  return (
    <footer className="footer container">
      <div className="footer__content">
        <a
          className="footer__rss-link"
          href="https://rs.school/react/"
          target="_blank"
          rel="noreferrer"
        >
          <svg className="footer__svg"></svg>
        </a>
        <div className="footer__gh">
          {developersEn.map((developer, index) => (
            <a
              key={index}
              className="footer__gh-link"
              href={`https://github.com/${developer.github}`}
              target="_blank"
              rel="noreferrer"
            >
              <svg className="footer__svg"></svg>
            </a>
          ))}
        </div>
        <div className="footer__year-link">
          <span>@Copyright {new Date().getFullYear()}</span>{' '}
          <a href={WELCOME_ROUTE}>GraphiQL</a>
        </div>
      </div>
    </footer>
  );
};

export default Footer;
