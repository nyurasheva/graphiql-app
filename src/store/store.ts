import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import userReducer from './slice/userSlice';
import requestReducer from './slice/requestSlice';

const reducer = combineReducers({
  user: userReducer,
  request: requestReducer,
});

const store = configureStore({
  reducer,
});

setupListeners(store.dispatch);

export default store;
