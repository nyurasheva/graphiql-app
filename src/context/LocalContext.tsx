import React, {
  createContext,
  useState,
  useEffect,
  useContext,
  ReactNode,
} from 'react';
import { LocalizationContextType, Translation } from './types';
import { translationsData } from './LocalData';

const LocalizationContext = createContext<LocalizationContextType | undefined>(
  undefined
);

type LocalizationProviderProps = {
  children: ReactNode;
};

export const LocalizationProvider: React.FC<LocalizationProviderProps> = ({
  children,
}) => {
  const [language, setLanguage] = useState<string>('en');
  const [translations, setTranslations] = useState<Translation>({});

  const changeLanguage = (newLanguage: string) => {
    setLanguage(newLanguage);
  };

  useEffect(() => {
    setTranslations(translationsData[language]);
  }, [language]);

  return (
    <LocalizationContext.Provider
      value={{ language, translations, changeLanguage }}
    >
      {children}
    </LocalizationContext.Provider>
  );
};

export const Localization = () => {
  const context = useContext(LocalizationContext);
  if (!context) {
    throw new Error('Localization must be used within a LocalizationProvider');
  }
  return context;
};
