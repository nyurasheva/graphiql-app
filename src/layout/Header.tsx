import { NavLink } from 'react-router-dom';
import { Logo } from '../components';
import {
  AUTH_ROUTE,
  MAIN_ROUTE,
  REGISTRATION_ROUTE,
  WELCOME_ROUTE,
} from '../constants/route';
import { LogOut } from '../components/Logout';
import { useAppSelector } from '../store/types';
import { useEffect, useMemo, useState } from 'react';
import { Localization } from '../context/LocalContext';
import { AuthState } from '../utils/AuthState';

const Header = () => {
  AuthState();
  const { isAuth } = useAppSelector((state) => state.user);
  const [isSticky, setIsSticky] = useState(false);
  const [menuVisible, setMenuVisible] = useState(false);
  const { translations, language, changeLanguage } = Localization();

  const handleLanguageChange = (newLanguage: string) => {
    changeLanguage(newLanguage);
  };

  const toggleMenu = () => {
    setMenuVisible(!menuVisible);
  };

  const closeMenu = () => {
    setMenuVisible(false);
  };

  const viewMainLink = useMemo(() => {
    {
      if (isAuth)
        return (
          <NavLink to={MAIN_ROUTE} className="link">
            {translations.main}
          </NavLink>
        );
    }
  }, [isAuth, translations.main]);

  const viewButtons = useMemo(() => {
    {
      if (isAuth) return <LogOut />;
      return (
        <>
          <NavLink to={REGISTRATION_ROUTE} onClick={closeMenu}>
            <button className="button button-second">
              {translations.registration}
            </button>
          </NavLink>
          <NavLink to={AUTH_ROUTE} onClick={closeMenu}>
            <button className="button">{translations.login}</button>
          </NavLink>
        </>
      );
    }
  }, [isAuth, translations.login, translations.registration]);

  useEffect(() => {
    const handleScroll = () => {
      setIsSticky(window.scrollY >= 100);
    };

    window.addEventListener('scroll', handleScroll);

    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, []);

  return (
    <header className={`header ${isSticky ? 'sticky' : ''}`}>
      <div className={`header__content container`}>
        <div className="header__logo">
          <Logo />
        </div>
        <div className="header__links">
          <NavLink to={WELCOME_ROUTE} className="link">
            {translations.welcome}
          </NavLink>
          {viewMainLink}
        </div>
        <div className="header__buttons">
          <span className="header__buttons-icon" onClick={toggleMenu}></span>
          <div
            className={`header__buttons-list ${menuVisible ? 'active' : ''}`}
          >
            {viewButtons}
          </div>
          <div className="language">
            <input
              type="checkbox"
              id="switch"
              className="language__input"
              defaultChecked={language === 'ru'}
              onChange={(e) => {
                if (e.target.checked) {
                  handleLanguageChange('ru');
                } else {
                  handleLanguageChange('en');
                }
              }}
            />
            <label htmlFor="switch" className="language__toggler">
              <span className="language__ru">{translations.ru}</span>
              <span className="language__en">{translations.en}</span>
            </label>
          </div>
        </div>
      </div>
    </header>
  );
};

export default Header;
