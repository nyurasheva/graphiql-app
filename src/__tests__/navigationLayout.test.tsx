import '@testing-library/jest-dom';
import { NavigationLayout } from '../layout';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../store/store';
import { BrowserRouter } from 'react-router-dom';
import { LocalizationProvider } from '../context/LocalContext';

describe('NavigationLayout component', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <LocalizationProvider>
        <Provider store={store}>
          <BrowserRouter>
            <NavigationLayout />
          </BrowserRouter>
        </Provider>
      </LocalizationProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
