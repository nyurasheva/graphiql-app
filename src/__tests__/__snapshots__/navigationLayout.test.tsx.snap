// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationLayout component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <header
      class="header "
    >
      <div
        class="header__content container"
      >
        <div
          class="header__logo"
        >
          <div
            class="logo"
          >
            <a
              aria-current="page"
              class="active"
              href="/"
              title=""
            >
              <img
                alt="logo"
                class="logo__image"
                src="test-file-stub"
                title=""
              />
              <span
                class="logo__span"
              >
                GraphiQL
              </span>
            </a>
          </div>
        </div>
        <div
          class="header__links"
        >
          <a
            aria-current="page"
            class="link active"
            href="/"
          >
            Welcome
          </a>
        </div>
        <div
          class="header__buttons"
        >
          <span
            class="header__buttons-icon"
          />
          <div
            class="header__buttons-list "
          >
            <a
              class=""
              href="/registration"
            >
              <button
                class="button button-second"
              >
                Registration
              </button>
            </a>
            <a
              class=""
              href="/login"
            >
              <button
                class="button"
              >
                Sign in
              </button>
            </a>
          </div>
          <div
            class="language"
          >
            <input
              class="language__input"
              id="switch"
              type="checkbox"
            />
            <label
              class="language__toggler"
              for="switch"
            >
              <span
                class="language__ru"
              />
              <span
                class="language__en"
              >
                En
              </span>
            </label>
          </div>
        </div>
      </div>
    </header>
    <main
      class="content"
    />
    <footer
      class="footer container"
    >
      <div
        class="footer__content"
      >
        <a
          class="footer__rss-link"
          href="https://rs.school/react/"
          rel="noreferrer"
          target="_blank"
        >
          <svg
            class="footer__svg"
          />
        </a>
        <div
          class="footer__gh"
        >
          <a
            class="footer__gh-link"
            href="https://github.com/nyurasheva"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              class="footer__svg"
            />
          </a>
          <a
            class="footer__gh-link"
            href="https://github.com/forsyterun"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              class="footer__svg"
            />
          </a>
          <a
            class="footer__gh-link"
            href="https://github.com/massaracsh7"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              class="footer__svg"
            />
          </a>
        </div>
        <div
          class="footer__year-link"
        >
          <span>
            @Copyright 2024
          </span>
           
          <a
            href="/"
          >
            GraphiQL
          </a>
        </div>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
